pkg/ccl/importccl/import_processor_test.go

```diff
@@ -172,10 +137,190 @@ func TestConverterFlushesBatches(t *testing.T) {
+func (r *errorReportingRowReceiver) ProducerDone() {}
+func (r *errorReportingRowReceiver) Types() []types.T {
+	return nil
+}
+
+// A do nothing BulkAdder implementation
+type doNothingBulkAdder struct{}
+
+var doNothingAdder storagebase.BulkAdder = &doNothingBulkAdder{}
+
+func (*doNothingBulkAdder) Add(_ context.Context, _ roachpb.Key, _ []byte) error { return nil }
+func (*doNothingBulkAdder) Flush(_ context.Context) error                        { return nil }
+func (*doNothingBulkAdder) IsEmpty() bool                                        { return true }
+func (*doNothingBulkAdder) CurrentBufferFill() float32                           { return 0 }
+func (*doNothingBulkAdder) GetSummary() roachpb.BulkOpSummary                    { return roachpb.BulkOpSummary{} }
+func (*doNothingBulkAdder) Close(_ context.Context)                              {}
+func (*doNothingBulkAdder) SetOnFlush(func())                                    {}
+
+func doNothingBulkAdderFactory(
+	_ context.Context, _ *client.DB, _ hlc.Timestamp, _ storagebase.BulkAdderOptions,
+) (storagebase.BulkAdder, error) {
+	return doNothingAdder, nil
+}
+
+var eofOffset uint64 = math.MaxUint64
+
+func TestImportIgnoresProcessedFiles(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	ctx := context.Background()
+	testServer, _, _ := serverutils.StartServer(t, base.TestServerArgs{})
+	defer testServer.Stopper().Stop(ctx)
+
+	evalCtx := tree.MakeTestingEvalContext(nil)
+	flowCtx := &execinfra.FlowCtx{
+		EvalCtx: &evalCtx,
+		Cfg: &execinfra.ServerConfig{
+			Settings:        testServer.ClusterSettings(),
+			ExternalStorage: externalStorageFactory,
+			BulkAdder:       doNothingBulkAdderFactory,
+		},
+	}
+
+	// In this test, we'll specify import files that do not exist, but mark those files fully processed.
+	// The converters should not attempt to even open these files (and if they do, we should report a test failure)
+	tests := []struct {
+		name         string
+		spec         testSpec
+		inputOffsets []uint64 // List of file ids that were fully processed
+	}{
+		{
+			"csv-two-invalid",
+			newTestSpec(t, roachpb.IOFileFormat_CSV, "__invalid__", "testdata/csv/data-0", "/_/missing/_"),
+			[]uint64{eofOffset, 0, eofOffset},
+		},
+		{
+			"csv-all-invalid",
+			newTestSpec(t, roachpb.IOFileFormat_CSV, "__invalid__", "../../&"),
+			[]uint64{eofOffset, eofOffset},
+		},
+		{
+			"csv-all-valid",
+			newTestSpec(t, roachpb.IOFileFormat_CSV, "testdata/csv/data-0"),
+			[]uint64{0},
+		},
+		{
+			"mysql-one-invalid",
+			newTestSpec(t, roachpb.IOFileFormat_Mysqldump, "testdata/mysqldump/simple.sql", "/_/missing/_"),
+			[]uint64{0, eofOffset},
+		},
+		{
+			"pgdump-one-input",
+			newTestSpec(t, roachpb.IOFileFormat_PgDump, "testdata/pgdump/simple.sql"),
+			[]uint64{0},
+		},
+	}
+
+	// Configures import spec to have appropriate input offsets set.
+	setInputOffsets := func(
+		t *testing.T, spec *execinfrapb.ReadImportDataSpec, offsets []uint64) *execinfrapb.ReadImportDataSpec {
```

Comment:
Nit: we usually put a trailing comma after last arg and then a newline so that the `) [returnTypes ] {` appears on its own _dedented_ line, breaking the arguments and body into visually distinct blocks. This is _usually_ done automatically by `crlfmt` but I think maybe your `func` literal didn’t get caught by it.

Reply:
Done